src/test/java/com/google/sps/servlets/NewRequestServletTest.java

```diff
@@ -95,23 +116,32 @@ public void testStartTimeAfterEndTime() throws IOException, ServletException {
     verify(response).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid time!");
   }
   /**
-   * Tests that when correct data is sent to the servlet, no errors are sent.
-   * This test will be added when Firebase Credentials will be set
+   * Tests that when correct data is sent to the servlet, no BAD_REQUEST errors are sent. 
    */
-//   @Test
-//   public void testValidData() throws IOException, ServletException {
-//     when(request.getParameter("delivery-date")).thenReturn("2020-09-26");
-//     when(request.getParameter("start-time")).thenReturn("05:15");
-//     when(request.getParameter("end-time")).thenReturn("15:10");
-//     when(request.getParameter("max-stops")).thenReturn("1");
-//     when(request.getParameter("timezone-offset")).thenReturn("180");
+  @Test
+  public void testValidData() throws IOException, ServletException {
+    when(request.getParameter("delivery-date")).thenReturn("2020-09-26");
+    when(request.getParameter("start-time")).thenReturn("05:15");
+    when(request.getParameter("end-time")).thenReturn("15:10");
+    when(request.getParameter("max-stops")).thenReturn("1");
+    when(request.getParameter("timezone-offset-minutes")).thenReturn("180");
+    when(request.getParameter("start-address")).thenReturn("277 Bedford Ave, Brooklyn, NY 11211, USA");
+
+    when(response.getWriter()).thenReturn(new PrintWriter(System.out));
 
-//     NewRequestServlet requestServlet = new NewRequestServlet();
-//     requestServlet.doPost(request, response);
-//     // the date is valid, thus "Please enter a valid date!" shouldn't be sent
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid date!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid time!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter between 1 and 25 stops!");
-//     //TODO[ak47na]: check that the delivery request is added to Mock datastore
-//   }
+    NewRequestServlet requestServlet = spy(new NewRequestServlet());
+    
+    requestServlet.doPost(request, response);
+    // The date is valid, thus "Please enter a valid date!" shouldn't be sent.
```

Comment:
In general, you don't need to do this. Tests should verify what should happen, not what shouldn't happen.

In your case, the only thing you should verify here, is that this returns 200 OK. The issue you're having here is that HttpServletRequest, which means you can't instantiate it directly. Appengine also doesn't seem to provide many helpers to make this easy (and mocking everything, like you're doing here seem painful and not super useful). Same with firebase from my quick searching (not sure if you've been more successful).

What you want to do in this case is move away as quickly as possible from using these helpers and set up your own classes/abstractions that you _can_ test (http://tutorials.jenkov.com/java-unit-testing/servlet-testing.html). Sadly you're going to have to change your code just to make it testable.

Here is a rough skeleton example:
```
// NewRequestServlet.java
@WebServlet("/new-delivery-request")
public class NewRequestServlet extends HttpServlet {
  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
    try {
      User user (getFirebaseUser().getUid()));
    } catch (UnauthenticatedException e) {
      response.SendError(SC_UNAUTHENTICATED, e.getMessage());
   }

    DeliverySlotManager slot_manager(DatastoreServiceFactory.getDatastoreService(), initializeFirebaseApp());
    // Create `create_slot_request` from `request` parameters.
    try {
      slot_manager.CreateSlot(create_slot_request, user);
    } catch (BadRequestException e) {
      response.SendError(SC_BAD_REQUEST, e.getMessage());
   }
  }
}

// Exceptions.java
class UnauthenticatedException extends Exception { ... }
class BadRequestException extends Exception { ... }

// User.java
class User {
  string uid;
}

// DeliverySlotManager.java
class CreateSlotRequest {
  string DeliveryDate;
  ...
}

class DeliverySlotManager {
  public DeliverySlotManager(Datastore datastore);
  public void CreateSlot(CreateSlotRequest req, User user) throws BadRequestException { // your code here };
}
```

The setup involves a bit more boilerplate, but it does make things super easy to test. You'll still not be able to easily test the servlet, but there's not much code in the servlet to test. Testing `DeliverySlotManager` is sufficient. This test then becomes:

```
@Test
public void testCreateSlotSuccessfully() throws IOException, ServletException {
  Request request;
  request.delivery-date = "..."
  deliverySlotManager.Createslot(request, User("foo"));
  // Query datastore and verify what is in there is what what you put in the request.
}

@Test
public void testCreateSlotInvalidDate() throws IOException, ServletException {
  Request request;
  request.delivery-date = "invalid";
  assertThrows(BadRequestException.class,
     () -> deliverySlotManager.Createslot(request, User("foo")));
}
```

Reply:
I added DeliverySlotManager, CreateSlotManager, BadRequestException, DataNotFoundException classes and the two new tests. I will also change how the start address is retrieved, but should that be in a new PR?